hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java

```diff
@@ -818,11 +835,17 @@ protected void applyQuerySettingsAndHints(Query query) {
 	@SuppressWarnings({"unchecked", "WeakerAccess", "StatementWithEmptyBody"})
 	protected void resultClassChecking(Class resultClass, org.hibernate.Query hqlQuery) {
 		// make sure the query is a select -> HHH-7192
-		final HQLQueryPlan queryPlan = getFactory().getQueryPlanCache().getHQLQueryPlan(
-				hqlQuery.getQueryString(),
-				false,
-				getLoadQueryInfluencers().getEnabledFilters()
-		);
+		HQLQueryPlan queryPlan = null;
+		if ( hqlQuery instanceof PlanAwareQueryImpl ) {
+			queryPlan = ( (PlanAwareQueryImpl) hqlQuery).plan;
+		}
+		if ( queryPlan == null ) {
+			queryPlan = getFactory().getQueryPlanCache().getHQLQueryPlan(
+					hqlQuery.getQueryString(),
+					false,
+					getLoadQueryInfluencers().getEnabledFilters()
+			);
+		}
```

Comment:
I retained the above code for the method is `protected` and it might have been used by our user. Let me know if it is overkill. Thanks.